UCT/API: Support interface query by operation and memory type

gleon99 opened the pull request:
## What
Support querying a UCT interface (`uct_iface_query`) by operation and memory type

## Why ?
Allow finer granularity in the information retrieved by querying the interface.
The change is breaking and not backwards compatible, but future fields can be added without breaking compatibility, as the structs are now refactored to hold pointers rather than contain nested data.

gleon99 commented on src/uct/api/uct.h:
```diff
@@ -850,105 +850,132 @@ typedef struct uct_ppn_bandwidth {
+    uct_atomic_attrs_t  *atomic32;   /**< Attributes for atomic operations */
+    uct_atomic_attrs_t  *atomic64;   /**< Attributes for atomic operations */
+
+    uint64_t            flags;       /**< Flags from @ref UCT_RESOURCE_IFACE_CAP */
```
Can move to `uct_iface_attr` (no need in per-memory type granularity).

gleon99 commented on src/uct/api/uct.h:
```diff
@@ -850,105 +850,132 @@ typedef struct uct_ppn_bandwidth {
+                                         ignored */
+    uint64_t            op_flags;   /**< Attributes for atomic-post operations */
+    uint64_t            fop_flags;  /**< Attributes for atomic-fetch operations */
+    double              overhead;   /**< Message overhead, seconds */
```
Maybe **only** these (performance) need to be per-memory type.

gleon99 commented on src/uct/api/uct.h:
```diff
@@ -850,105 +850,132 @@ typedef struct uct_ppn_bandwidth {
+                                             using bits from @ref uct_op_attrs_field.
+                                             Fields not specified in this mask will be
+                                             ignored */
+    size_t              min_recv;       /**< Minimal allowed length of posted
```
This + `max_ountstanding` - put in a separate struct (tag matching).

gleon99 commented on src/uct/api/uct.h:
```diff
@@ -850,105 +850,149 @@ typedef struct uct_ppn_bandwidth {
+    uint64_t field_mask;
+    double              overhead;   /**< Message overhead, seconds */
+    uct_ppn_bandwidth_t bandwidth;  /**< Bandwidth model */
+} uct_performance_attrs_t;
```
Guideline: in slow path funcs, have the params passed as struct members.

gleon99 commented on src/uct/base/uct_iface.c:
```diff
@@ -178,6 +178,30 @@ ucs_status_t uct_iface_query(uct_iface_h iface, uct_iface_attr_t *iface_attr)
 }
 
+ucs_status_t uct_iface_query_performance(uct_iface_h iface, uct_performance_attrs_t *performance,
+                                         ucs_memory_type_t first, ucs_memory_type_t second,
```
Have an input struct here i.e `perf_query_params`.

gleon99 commented on src/uct/base/uct_iface.c:
```diff
@@ -178,6 +178,30 @@ ucs_status_t uct_iface_query(uct_iface_h iface, uct_iface_attr_t *iface_attr)
+ucs_status_t uct_iface_query_performance(uct_iface_h iface, uct_performance_attrs_t *performance,
+                                         ucs_memory_type_t first, ucs_memory_type_t second,
+                                         uct_iface_cap_field_t op, uct_performance_attrs_field_t field_mask)
+{
```
Add test

gleon99 commented on src/uct/api/uct.h:
```diff
@@ -346,6 +346,34 @@ typedef enum uct_atomic_op {
+ *
+ * This enumeration defines all available UCT operations.
+ */
+typedef enum uct_op {
```
Some enum which has all ops in a single type was needed. First I considered using `uct_atomic_op_t` together with `UCT_IFACE_FLAG_....` but that didn't seem feasible

gleon99 replied:
Change to match `uct_ep_operation_t`

gleon99 commented:
/azp run

gleon99 commented:
/azp run

yosefe commented on src/uct/api/uct.h:
```diff
@@ -1062,6 +1090,73 @@ struct uct_iface_params {
+ */
+enum uct_perf_attr_field {
+    /** Enables @ref uct_perf_attr_t::first_memory_type */
+    UCT_PERF_ATTR_FIRST_MEMORY_TYPE  = UCS_BIT(0),
```
LOCAL_MEMORY_TYPE

yosefe commented on src/uct/api/uct.h:
```diff
@@ -1062,6 +1090,73 @@ struct uct_iface_params {
+    UCT_PERF_ATTR_FIRST_MEMORY_TYPE  = UCS_BIT(0),
+
+    /** Enables @ref uct_perf_attr_t::second_memory_type */
+    UCT_PERF_ATTR_SECOND_MEMORY_TYPE = UCS_BIT(1),
```
REMOTE_MEMORY_TYPE

yosefe commented on src/uct/api/uct.h:
```diff
@@ -1062,6 +1090,73 @@ struct uct_iface_params {
+    UCT_PERF_ATTR_SECOND_MEMORY_TYPE = UCS_BIT(1),
+
+    /** Enables @ref uct_perf_attr_t::operation */
+    UCT_PERF_ATTR_OPERATION          = UCS_BIT(2),
```
UCT_IFACE_PERF_ATTR_FIELD_OPERATION

yosefe commented on src/uct/api/uct.h:
```diff
@@ -1062,6 +1090,73 @@ struct uct_iface_params {
+    /**
+     * Operation to report performance for.
+     */
+    uct_op_t            operation;
```
uct_ep_operation_t

yosefe commented on src/uct/api/uct.h:
```diff
@@ -1781,6 +1876,17 @@ void uct_iface_close(uct_iface_h iface);
 
+/**
+ * @ingroup UCT_RESOURCE
+ * @brief Get interface performance attributes, by memory types and operation
```
add details: that user should fill op, memtype, etc and then will have overhead&bandwidth as the output

yosefe commented on src/uct/base/uct_iface.c:
```diff
@@ -178,6 +178,27 @@ ucs_status_t uct_iface_query(uct_iface_h iface, uct_iface_attr_t *iface_attr)
+        return status;
+    }
+
+    if (perf_attr->field_mask & UCT_PERF_ATTR_BANDWIDTH) {
```
add a comment that by default the performance is assumed to be the same for all operations

gleon99 commented:
@yosefe , CR comments fixed.

gleon99 commented on src/uct/api/uct.h:
```diff
@@ -1062,6 +1090,73 @@ struct uct_iface_params {
+    UCT_PERF_ATTR_FIELD_REMOTE_MEMORY_TYPE = UCS_BIT(1),
+
+    /** Enables @ref uct_perf_attr_t::operation */
+    UCT_PERF_ATTR_FIELD_OPERATION          = UCS_BIT(2),
```
Make op the first one (also in struct & everywhere)

gleon99 commented on src/uct/api/uct.h:
```diff
@@ -1062,6 +1090,73 @@ struct uct_iface_params {
+ *
+ * This structure should be allocated and initialized by the user
+ */
+typedef struct {
```
Explain in each field whether it's in/out.

gleon99 commented on test/gtest/uct/test_uct_query.cc:
```diff
@@ -0,0 +1,41 @@
+    EXPECT_EQ(iface_attr.bandwidth.shared, perf_attr.bandwidth.shared);
+    EXPECT_EQ(iface_attr.overhead, perf_attr.overhead);
+}
+UCT_INSTANTIATE_TEST_CASE(test_uct_query)
```
Add blank line

gleon99 commented on test/gtest/uct/test_uct_query.cc:
```diff
@@ -0,0 +1,41 @@
+    status = uct_iface_query_perf(m_entities.at(0).iface(), &perf_attr);
+    EXPECT_EQ(status, UCS_OK);
+
+    status = uct_iface_query(m_entities.at(0).iface(), &iface_attr);
```
Use `sender()`.

gleon99 commented on test/gtest/uct/test_uct_query.cc:
```diff
@@ -0,0 +1,41 @@
+    {
+        uct_test::init();
+
+        entity *receiver = uct_test::create_entity(0);
```
Remove var.

gleon99 commented:
@yosefe , CR comments fixed.

yosefe commented on Makefile.am:
```diff
@@ -13,7 +13,7 @@
 ACLOCAL_AMFLAGS = -I config/m4
 
-noinst_HEADERS = src/uct/api/uct.h src/uct/api/uct_def.h src/uct/api/tl.h
+noinst_HEADERS = src/uct/api/uct.h src/uct/api/v2/uct.h src/uct/api/uct_def.h src/uct/api/tl.h
```
pls arrange each H file in a separate line with `\`
see SUBDIRS below as example

yosefe commented on src/uct/api/v2/uct.h:
```diff
@@ -0,0 +1,144 @@
+ * @file        uct.h
+ * @date        2014-2021
+ * @copyright   NVIDIA Corporation. All rights reserved.
+ * @copyright   Mellanox Technologies Ltd. All rights reserved.
+ * @copyright   Oak Ridge National Laboratory. All rights received.
+ * @copyright   Advanced Micro Devices, Inc. All rights received.
```
can remove those

gleon99 commented:
@yosefe , fixed.

yosefe commented on src/uct/api/v2/uct_v2.h:
```diff
@@ -54,7 +50,9 @@ typedef enum uct_ep_operation {
     UCT_OP_ATOMIC_XOR,   /**< Atomic xor */
     UCT_OP_ATOMIC_SWAP,  /**< Atomic swap */
-    UCT_OP_ATOMIC_CSWAP  /**< Atomic compare-and-swap */
+    UCT_OP_ATOMIC_CSWAP, /**< Atomic compare-and-swap */
```
remove UCT_OP_ATOMIC which are not post/fetch

yosefe commented on src/uct/api/v2/uct_v2.h:
```diff
@@ -32,25 +32,6 @@ BEGIN_C_DECLS
-    UCT_OP_ATOMIC_ADD,   /**< Atomic add */
-    UCT_OP_ATOMIC_AND,   /**< Atomic and */
-    UCT_OP_ATOMIC_OR,    /**< Atomic or */
-    UCT_OP_ATOMIC_XOR,   /**< Atomic xor */
```
too much removed

gleon99 commented:
@yosefe reverted.

yosefe commented on src/uct/api/v2/uct_v2.h:
```diff
@@ -0,0 +1,138 @@
+/**
+ * @file        uct.h
+ * @date        2021
```
pls add just Mellanox copyright

yosefe commented on src/uct/api/v2/uct_v2.h:
```diff
@@ -0,0 +1,138 @@
+
+BEGIN_C_DECLS
+
+/** @file uct.h */
```
update fname

yosefe replied:
update fname

yosefe commented on test/gtest/uct/test_uct_query.cc:
```diff
@@ -0,0 +1,39 @@
+    uct_perf_attr_t perf_attr;
+    ucs_status_t status;
+
+    perf_attr.field_mask = UCT_PERF_ATTR_FIELD_BANDWIDTH |
+                           UCT_PERF_ATTR_FIELD_OVERHEAD;
+    status               = uct_iface_query_perf(sender().iface(), &perf_attr);
+    EXPECT_EQ(status, UCS_OK);
```
pass operation and memory types

gleon99 commented:
@yosefe , fixed.

yosefe commented on test/gtest/uct/test_uct_query.cc:
```diff
@@ -24,9 +24,13 @@ UCS_TEST_P(test_uct_query, query_perf)
-    perf_attr.field_mask = UCT_PERF_ATTR_FIELD_BANDWIDTH |
-                           UCT_PERF_ATTR_FIELD_OVERHEAD;
-    status               = uct_iface_query_perf(sender().iface(), &perf_attr);
+    perf_attr.field_mask         = UCT_PERF_ATTR_FIELD_BANDWIDTH |
```
update field mask

yosefe commented:
@shamisp @tonycurtis pls take a look

tonycurtis reviewed (changes requested):
uct_v2.h:25-26: not sure what "a concept" means here.  Is it something like "describes how the communication resource works"?
83: should -> must ?
113: "message overhead time, in seconds" seems clearer

gleon99 commented:
Lines 83, 113 - fixed.
uct_v2.h:25-26 - it is copied from `uct.h`.

gleon99 commented:
@tonycurtis - does that look OK?

gleon99 commented:
@shamisp , can look please?

shamisp commented on test/gtest/uct/test_uct_query.cc:
```diff
@@ -0,0 +1,46 @@
+ * See file LICENSE for terms.
+ */
+
+extern "C" {
```
@yosefe can we change the file name to `test_uct_v2_query` or move this to `uct/v2/` folder ?

yosefe replied:
i think we actually want to transition tests to use the new private APIs, and test old APIs in something like "test_uct_compat.cc"

shamisp replied:
well, then it makes even more sense to add some prefix for now. Once the API stable we can move  uct_*.cc -> uct_*_compat.cc and uct_*_compat.cc -> uct_*.cc

gleon99 replied:
Let's leave it `test_uct_query.cc`, as discussed with @yosefe .

yosefe replied:
this way, after we decide to switch to new API, will not need to change tests

shamisp replied:
@yosefe and I discussed offline that it makes sense as for now to move the tests to `uct/v2`. Later we can rearrange this

yosefe replied:
+1
@shamisp can we work on the _v2 renaming macros in another PR? this PR does not change an existing function so there is no name collision

shamisp replied:
@yosefe yes, but I was commenting on the test location (aka move it to v2 subfolder). I agree with _v2 C macro development in a separate PR.

yosefe replied:
@shamisp +1 for the test location

shamisp commented on src/uct/api/v2/uct_v2.h:
```diff
@@ -0,0 +1,139 @@
+ * @param [inout] perf_attr Filled with performance attributes.
+ */
+ucs_status_t
+uct_iface_query_perf(uct_iface_h iface, uct_perf_attr_t *perf_attr);
```
@yosefe For consistency reasons, shell we name it uct_iface_query_perf_v2 since the other APIs will have the suffix ?

shamisp replied:
It is not very clear to me how the mask is used in the context of `uct_iface_query_perf()`. Shell the user specify `field_mask` for the fields that users wants to query.

yosefe replied:
> @yosefe For consistency reasons, shell we name it uct_iface_query_perf_v2 since the other APIs will have the suffix ?

IMO, for new functions, no need to have _v2 suffix. It will make transitions easier.

shamisp replied:
@yosefe - IMO, it will be very confusing.

gleon99 replied:
As discussed with @yosefe , we'll stick with `uct_iface_query_perf` (other API is `uct_ib_iface_is_roce_v2` which is irrelevant).

yosefe replied:
If we add _v2 suffix to everything now, even when not really needed to avoid name collision, will require extra steps when making these APIs default. We don't want to be stuck for ever with _v2 in the API...

yosefe commented:
@shamisp the test file was moved to /v2 subdir, is it ok now?

shamisp commented:
👍

gleon99 commented:
Squashed. Renamed `uct_iface_query_perf` to `uct_iface_estimate_perf`, as discussed in https://github.com/openucx/ucx/pull/6403#discussion_r586463750.
